Lf 3363 repeat crop plan repeat crop plan endpoint

Duncan-Brain opened the pull request:
**Description**

By sending a "template" (or source) management plan id, an array of start dates, and repeatDetails this endpoint will respond with an array of [ {management_plan: mp, tasks: [tsks]}, {..}].

Originally I had hoped to reuse the addManagementPlan endpoint. But since the decisions about WHICH tasks to create are done on that controller (instead of a pre-formed post request from frontend) I decided not to remove (or reposition) that logic. 

Design decisions:
- Keep management controller simple and offload the complex graph generation to copyCropPlan.js.
- Add new utils for generic utils
- Added custom static method to Models in the template
  - went with an 'edit', 'omit', 'keep' syntax as a reminder about what should be dropped, changed, untouched when making templates
  - a point was made that a simple bool would suffice here so I am open to changing this syntax
  - keeping this function in Models helps developers be less likely that a change to the model will go unnoticed on the template
  - locations were labeled as 'edit' and set explicitly as a nod to upcoming work to allow configuration
- lodash was kept but individual functions were loaded instead of the whole package.
 

Things I am not confident on currently:
- It feels like I am not doing much validation on the received data, should I send more bad requests for instances where dates are prior to template management plan?
- The route access validation: is hasFarmAccess(management_plan_id) enough?
- return object may contain some more data than necessary, should I remove null data?
- naming things is hard - my jest test feels exceptionally poorly named


Jira link: https://lite-farm.atlassian.net/browse/LF-3363

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Passes test case
- [x] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [x] Other (please explain):
Jest test available: `npm run test -- managementPlan.test.js`

Postman on /repeat_plan endpoint:
Example Postman body for /repeat_plan endpoint: 
```
{
    "management_plan_id": "1958",
    "startDates": ["2023-07-20T00:00:00.000","2023-07-27T00:00:00.000", "2023-08-03T00:00:00.000","2023-08-10T00:00:00.000"],
    "repeatDetails": {
        "planName": "The best plan"
    }
}

```
**Checklist:**

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [ ] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

iperdomo commented:
👏🏾  I was able to hit the endpoint `/management_plan/repeat_plan` with the testing sample dataset 
![2023-07-25_988x538](https://github.com/LiteFarmOrg/LiteFarm/assets/178474/4efbe354-740b-41f9-857b-9e22be944e98)

```
curl -v 'http://localhost:5001/management_plan/repeat_plan' -H 'Accept: application/json' -H 'user_id: cfee062e-09eb-11ee-8ab9-7ac3b12dfaeb' -H 'farm_id: 4d39c914-09ed-11ee-8209-7ac3b12dfaeb' -H 'Authorization: Bearer <token>' -H 'Content-Type: application/json' -d '{"management_plan_id": 2, "startDates":["2022-09-01T00:00:00.000"], "repeatDetails": {"planName": "Testing endpoint"}}'

```
Result
```
< HTTP/1.1 201 Created
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization
< Content-Type: application/json; charset=utf-8
< Content-Length: 5459
< ETag: W/"1553-kwOKAzM+2FiEIzx9rndz82XFSGE"
< Date: Tue, 25 Jul 2023 11:44:37 GMT
< Connection: keep-alive
< Keep-Alive: timeout=5
< 
[
  {
    "management_plan": {
      "crop_variety_id": "87d678d0-09fa-11ee-bd9e-7ac3b12dfaeb",
      "notes": "",
      "name": "Testing endpoint",
      "crop_management_plan": {
        "estimated_revenue": null,
        "already_in_ground": false,
        "is_seed": true,
        "needs_transplant": true,
        "for_cover": false,
        "is_wild": null,
        "estimated_yield": 60,
        "estimated_yield_unit": "kg",
        "estimated_price_per_mass": null,
        "estimated_price_per_mass_unit": "kg",
        "seed_date": "2022-09-01T00:00:00.000",
        "plant_date": null,
        "germination_date": "2022-09-05T00:00:00.000",
        "transplant_date": "2022-10-12T00:00:00.000",
        "harvest_date": "2022-11-29T00:00:00.000",
        "termination_date": null,
        "planting_management_plans": [
          {
            "is_final_planting_management_plan": false,
            "planting_method": "CONTAINER_METHOD",
            "is_planting_method_known": null,
            "estimated_seeds": 0.005,
            "estimated_seeds_unit": "kg",
            "pin_coordinate": null,
            "notes": null,
            "planting_task_type": "PLANT_TASK",
            "planting_management_plan_id": "9df343f5-0f47-4233-ab61-1713a96df810",
            "location_id": "dc4e4516-09f3-11ee-8a72-7ac3b12dfaeb",
            "transplant_task": null,
            "bed_method": null,
            "container_method": {
              "in_ground": true,
              "plant_spacing": 0.05,
              "plant_spacing_unit": "cm",
              "total_plants": 500,
              "number_of_containers": null,
              "plants_per_container": null,
              "planting_depth": 0.05,
              "planting_depth_unit": "cm",
              "planting_soil": null,
              "container_type": null,
              "planting_management_plan_id": "9df343f5-0f47-4233-ab61-1713a96df810"
            },
            "broadcast_method": null,
            "row_method": null,
            "management_plan_id": 3
          },
          {
            "is_final_planting_management_plan": true,
            "planting_method": "BED_METHOD",
            "is_planting_method_known": null,
            "estimated_seeds": 0.022222222222,
            "estimated_seeds_unit": "kg",
            "pin_coordinate": null,
            "notes": null,
            "planting_task_type": "TRANSPLANT_TASK",
            "planting_management_plan_id": "f5475207-5179-43cd-b776-f29931bcc6ce",
            "location_id": "dc4e4516-09f3-11ee-8a72-7ac3b12dfaeb",
            "plant_task": null,
            "bed_method": {
              "number_of_beds": 10,
              "number_of_rows_in_bed": 10,
              "plant_spacing": 0.225,
              "plant_spacing_unit": "cm",
              "bed_length": 5,
              "bed_length_unit": "m",
              "planting_depth": 0.01,
              "planting_depth_unit": "cm",
              "bed_width": null,
              "bed_width_unit": "cm",
              "bed_spacing": null,
              "bed_spacing_unit": "cm",
              "specify_beds": null,
              "planting_management_plan_id": "f5475207-5179-43cd-b776-f29931bcc6ce"
            },
            "container_method": null,
            "broadcast_method": null,
            "row_method": null,
            "management_plan_id": 3
          }
        ],
        "management_plan_id": 3
      },
      "repetition_number": 1,
      "management_plan_group_id": "2fc80250-3b0c-4ce1-8af5-d6bdd04031ba",
      "management_plan_id": 3,
      "management_plan_group": {
        "management_plan_group_id": "2fc80250-3b0c-4ce1-8af5-d6bdd04031ba",
        "repetition_count": 1
      }
    },
    "tasks": [
      {
        "notes": null,
        "task_type_id": 5,
        "coordinates": null,
        "locationTasks": [],
        "due_date": "2022-09-01T00:00:00.000",
        "owner_user_id": "cfee062e-09eb-11ee-8ab9-7ac3b12dfaeb",
        "assignee_user_id": null,
        "wage_at_moment": null,
        "task_id": 19,
        "plant_task": {
          "planting_management_plan_id": "9df343f5-0f47-4233-ab61-1713a96df810",
          "task_id": 19
        }
      },
      {
        "notes": null,
        "task_type_id": 19,
        "coordinates": null,
        "locationTasks": [],
        "due_date": "2022-10-12T00:00:00.000",
        "owner_user_id": "cfee062e-09eb-11ee-8ab9-7ac3b12dfaeb",
        "assignee_user_id": null,
        "wage_at_moment": null,
        "task_id": 20,
        "transplant_task": {
          "planting_management_plan_id": "f5475207-5179-43cd-b776-f29931bcc6ce",
          "prev_planting_management_plan_id": "9df343f5-0f47-4233-ab61-1713a96df810",
          "task_id": 20
        }
      },
      {
        "notes": null,
        "task_type_id": 8,
        "coordinates": null,
        "due_date": "2022-11-29T00:00:00.000",
        "owner_user_id": "cfee062e-09eb-11ee-8ab9-7ac3b12dfaeb",
        "assignee_user_id": null,
        "wage_at_moment": null,
        "pest_control_task": null,
        "irrigation_task": null,
        "scouting_task": null,
        "soil_task": null,
        "soil_amendment_task": null,
        "field_work_task": null,
        "harvest_task": {
          "projected_quantity": null,
          "projected_quantity_unit": "kg",
          "harvest_everything": true,
          "task_id": 21
        },
        "cleaning_task": null,
        "task_id": 21
      }
    ]
  }
```

iperdomo commented:
Note:
* We need to start thinking in prioritizing https://lite-farm.atlassian.net/browse/LF-3419 . I _wasted_ quite some time trying to debug an `undefined` issue just to realize that I was sending a wrong property name: `repetitionDetails` instead of `repeatDetails`

Duncan-Brain commented:
Sorry about that @iperdomo! I have updated the PR text. 

Agreed on LF-3419.

SayakaOno commented on packages/api/src/controllers/managementPlanController.js:
```diff
@@ -14,23 +14,106 @@
+  repeatManagementPlan() {
+    return async (req, res) => {
+      const trx = await transaction.start(Model.knex());
+      const { startDates, management_plan_id, repeatDetails } = req.body;
```
What do you think `repeatDeails` should look like? The user input will be:
```
{
    crop_plan_name: 'Repetitions of Apricot plan template',
    plan_start_date: '2023-07-25',
    repeat_frequency: 1,
    repeat_interval: { value: 'year', label: 'year(s)' },
    days_of_week: ['Tuesday'],
    finish: 'after',
    after_occurrences: '5',
  }
```
The only thing that matters for now is if I add `planName` or change `crop_plan_name` to `planName`.

Duncan-Brain replied:
Whatever works for you ! I have no preference.  I think I was working off of that last slack conversation about the shape with Joyce.

It probably make sense for me to change so your names stay consistent snake case.

SayakaOno replied:
I think it's more natural to leave the user input as is, too.
Could you use `crop_plan_name`?

Duncan-Brain replied:
got it, will do!

iperdomo commented on packages/api/src/controllers/managementPlanController.js:
```diff
@@ -14,23 +14,109 @@
+  repeatManagementPlan() {
+    return async (req, res) => {
+      const trx = await transaction.start(Model.knex());
+      const { startDates, management_plan_id, repeatDetails } = req.body;
```
There is a mixture of `snake_case` and `camelCase` in the property names, we should try to decide on casing.
e.g.
```
src/controllers/sensorController.js
534:      const { start_time, end_time } = req.body;
548:      const { locationIds = [], readingTypes = [], endUnixTime = 0, startUnixTime = 0 } = req.body;
589:      const { location_id } = req.body;
```
If we decide on `snake_case`, we should change to `start_dates`, `repeat_details`, `management_plan_id`

Duncan-Brain replied:
I can set the rest to snake case I think I was basing it off an earlier discussion. If @SayakaOno and @kathyavini can adjust to snake case on their parts as needed too?

kathyavini replied:
I don't think my part is relevant as it goes through LF-3378, so it's all good by me 👍

Duncan-Brain replied:
now in snake case

iperdomo commented on packages/api/src/models/cropManagementPlanModel.js:
```diff
@@ -86,6 +86,34 @@ class CropManagementPlanModel extends Model {
+      transplant_date: 'edit',
+      harvest_date: 'edit',
+      termination_date: 'edit',
+      //TODO: convert already in ground to edit when planting method is known
```
This `TODO` is valid? If doesn't need to addressed, let's remove it or create a different ticket to tackle it later

iperdomo commented on packages/api/src/models/plantingManagementPlanModel.js:
```diff
@@ -39,7 +39,7 @@ class plantingManagementPlanModel extends Model {
         planting_management_plan_id: { type: 'string' },
         management_plan_id: { type: 'integer' },
-        is_final_planting_management_plan: { type: 'boolean' },
+        is_final_planting_management_plan: { type: ['boolean', 'null'] }, //UNDO: null change transplant task violates this rule somehow
```
`UNDO` still valid?

iperdomo commented on packages/api/src/models/plantingManagementPlanModel.js:
```diff
@@ -135,6 +135,34 @@ class plantingManagementPlanModel extends Model {
+      // jsonSchema()
+      planting_management_plan_id: 'edit',
+      management_plan_id: 'omit',
+      is_final_planting_management_plan: 'keep', //UNDO: null change transplant task violates this rule somehow
```
Idem, `UNDO`?

iperdomo commented on packages/api/src/util/copyCropPlan.js:
```diff
@@ -0,0 +1,639 @@
+            ..._omit(plan, getPropertiesToDelete(PlantingManagementPlanModel)),
+            planting_management_plan_id:
+              newPlantingManagementPlanUUIDs[plan.planting_management_plan_id],
+            location_id: plan.location_id, //TODO: allow location changing
```
`TODO` still valid?

iperdomo reviewed (changes requested):
Let's check the `TODO` and `UNDO` comments. If still valid and not need to be addressed, please create a ticket and remove the comments.

SayakaOno commented on packages/api/src/util/copyCropPlan.js:
```diff
@@ -0,0 +1,639 @@
+    : [];
+
+  // If there are properties in the template mapping schema, filter those marked for deletion ('omit' or 'edit').
+  return propertyAndRelationKeys.length
+    ? propertyAndRelationKeys.filter((key) =>
+        ['omit', 'edit'].includes(Model.templateMappingSchema[key]),
+      )
+    : [];
```
No need to check the length because `[]` will be returned for `[].filter()`.
(we could just make it short)

SayakaOno commented on packages/api/src/util/copyCropPlan.js:
```diff
@@ -0,0 +1,639 @@
+    // Check if there are plant tasks, transplant tasks or management tasks
+    // and add their complete_date or due_date to the dates array.
+    if (plan.plant_task) {
+      plan.plant_task.task.complete_date
+        ? dates.push(plan.plant_task.task.complete_date)
+        : dates.push(plan.plant_task.task.due_date);
```
Just as another option to do the same thing:
```
dates.push(plan.plant_task.task.complete_date || plan.plant_task.task.due_date)
```

Duncan-Brain replied:
I am still learning the best use of ternary vs null coalescing vs || &&. But I think an issue is cause by switching to the || here.

The test fails when the change is made indicating that the incorrect date may be pushed. I think because it is doing an array action it will "successfully" push a null complete_date.

SayakaOno commented on packages/api/src/util/copyCropPlan.js:
```diff
@@ -0,0 +1,639 @@
+ */
+export const getDatesFromManagementPlanGraph = (managementPlanGraph) => {
+  const dates = [];
+  managementPlanGraph.crop_management_plan.planting_management_plans.forEach((plan) => {
```
It seems that `plant_task` could be destructured here so that we don't have to repeat `plan.`.

SayakaOno commented on packages/api/src/util/copyCropPlan.js:
```diff
@@ -0,0 +1,639 @@
+};
+
+/**
+ * Used for jest tests only generates a management plan group graph based on the provided parameters.
```
@iperdomo Where are functions for tests usually placed? I am wondering if we should have a util folder under `tests`.

Duncan-Brain replied:
Good point, I didn't look to hard for this... But also I do kind of like the idea of all specific feature utils in one place.

iperdomo replied:
Agree with @SayakaOno test utility functions should be under `tests/utils/*`

Duncan-Brain replied:
@iperdomo @SayakaOno Looking at this further are you sure this is desirable? Utilities are utilities no matter where they are?

Doing this results in the following:

managementPlanTest imports regular copyUtils containing these:
  
![Screenshot 2023-07-27 at 8 34 38 AM](https://github.com/LiteFarmOrg/LiteFarm/assets/30075973/0875f79d-039c-4ed8-8e98-60bc0b41d792)

managementPlanTest now also imports the test utils which imports many of the same imports:
![Screenshot 2023-07-27 at 8 33 02 AM](https://github.com/LiteFarmOrg/LiteFarm/assets/30075973/25a6ea4e-46de-4a45-80f7-db7e3454f120)

test utils also actually needs the regular copy utils to function too (first picture). 

...So we are possibly duplicating these same imports 3 times

SayakaOno reviewed (commented):
Great job!!!
I added some comments on possible improvements, but you can ignore them for now :)
The license is missing in new files!

Duncan-Brain commented on packages/api/src/controllers/managementPlanController.js:
```diff
@@ -14,23 +14,109 @@
+      const { startDates, management_plan_id, repeatDetails } = req.body;
+      try {
+        if (!startDates?.length > 0 || !management_plan_id || !repeatDetails?.crop_plan_name) {
+          throw 'Insufficient details to copy crop plan';
```
When I try replacing this line with that syntax it does not work as expected.

Duncan-Brain commented on packages/api/tests/managementPlan.test.js:
```diff
@@ -1084,5 +1180,29 @@ describe('ManagementPlan Tests', () => {
     });
+
+    test('should copy an existing management plan with transplant', async (done) => {
+      let response;
```
Thanks! I am surprise linting did not pick it up.

Duncan-Brain commented on packages/api/tests/managementPlan.test.js:
```diff
@@ -1084,5 +1180,29 @@ describe('ManagementPlan Tests', () => {
+        userFarm,
+        async (err, res) => {
+          if (err) {
+            throw err;
```
I was finding that the test was falsely passing and I think this helps. I believe similar to : https://stackoverflow.com/questions/66278618/chai-expect-doesnt-fail-the-test-if-it-is-inside-a-callback

Duncan-Brain commented on packages/api/tests/managementPlan.test.js:
```diff
@@ -987,6 +1020,69 @@ describe('ManagementPlan Tests', () => {
+        .first();
+    }
+
+    async function expectManagementGroupPosted(res, management_plan_id) {
```
I think I was thinking I should use the original id instead of something passed around.. but ya request body is the same and I agree. Adding it in.

Duncan-Brain commented:
@iperdomo I removed the TODO/UNDO's but left some of the context in place with Jira ticket numbers -- it is kind of hard for me to separate useful context as a comment to completely delete that "bookmark". Some were deleted.

Please see the following issues found during this ticket:
[LF-3450](https://lite-farm.atlassian.net/browse/LF-3450)
[LF-3458](https://lite-farm.atlassian.net/browse/LF-3458)
[LF-3459](https://lite-farm.atlassian.net/browse/LF-3459)
[LF-3460](https://lite-farm.atlassian.net/browse/LF-3460)
[LF-3470](https://lite-farm.atlassian.net/browse/LF-3470)
[LF-3471](https://lite-farm.atlassian.net/browse/LF-3471)
[LF-3472](https://lite-farm.atlassian.net/browse/LF-3472)
[LF-3473](https://lite-farm.atlassian.net/browse/LF-3473)

@SayakaOno 

I fixed several of those comments -- others did not quite work for me. Thanks for finding those! I think I left comments on the ones I could get working.

kathyavini commented on packages/api/src/controllers/managementPlanController.js:
```diff
@@ -14,23 +14,107 @@
+            'crop_management_plan.[planting_management_plans.[managementTasks.[task.[pest_control_task, irrigation_task, scouting_task, soil_task, soil_amendment_task, field_work_task, harvest_task, cleaning_task, locationTasks]], plant_task.[task.[locationTasks]], transplant_task.[task.[locationTasks]], bed_method, container_method, broadcast_method, row_method]]',
+          )
+          .first();
+
```
This is missing a filter on deleted tasks which is needed according to ticket scope. But I'm sure why the created graph doesn't have the deleted property within the task objects. Do you know?

iperdomo reviewed (changes requested):
As per comment by @kathyavini we should exclude deleted tasks (those with deleted=true)

iperdomo reviewed (approved):
As agreed with the team, we're going to merge the PR and address the deleted tasks in a different PR
